### PR TITLE
Fix video freezes when switching to a layer that does not exist anymore

### PR DESCRIPTION
**Pull request by kekkokk**
There are situations in which the quality handler took lot of time to recognise a spatial layer doesn't exist anymore. with this trick we can deliver a better user experience switching down the resolution if the inter frame interval is > 300 ms (or whatever we decide to set)

[] It needs and includes Unit Tests
[] It includes documentation for these changes in `/doc`.

### CODE REVIEW

File: erizo/src/erizo/rtp/QualityFilterHandler.cpp [hunk 1]
```diff
@@ -120,6 +127,23 @@ void QualityFilterHandler::write(Context *ctx, std::shared_ptr<DataPacket> packe
   if (is_scalable_ && !chead->isRtcp() && enabled_ && packet->type == VIDEO_PACKET) {
     RtpHeader *rtp_header = reinterpret_cast<RtpHeader*>(packet->data);
 
+    if (cooldown && clock::now() > cooldown_target) {
+    	cooldown = false;
+    	ELOG_DEBUG("End cooldown")
```
Comment (jcague) [inline erizo/src/erizo/rtp/QualityFilterHandler.cpp]: remove this log or add more info please

File: erizo/src/erizo/rtp/QualityFilterHandler.cpp [hunk 2]
```diff
@@ -120,6 +127,23 @@ void QualityFilterHandler::write(Context *ctx, std::shared_ptr<DataPacket> packe
...
+    }
+
+    // Force switch to lower resolution if the packet frequency is > kSwitchDownTimeout (ms)
+    if (!cooldown && clock::now() - last_sent_packet > kSwitchDownTimeout) {
+    	ELOG_DEBUG("Switching to lower quality");
```
Comment (jcague) [inline erizo/src/erizo/rtp/QualityFilterHandler.cpp]: remove this log or add more info please

File: erizo/src/erizo/rtp/QualityFilterHandler.cpp [hunk 3]
```diff
@@ -50,8 +52,11 @@ void QualityFilterHandler::read(Context *ctx, std::shared_ptr<DataPacket> packet
 }
 
 void QualityFilterHandler::checkLayers() {
+  if (cooldown) {
+    return;
+  }
   int new_spatial_layer = quality_manager_->getSpatialLayer();
-  if (new_spatial_layer != target_spatial_layer_) {
+  if (new_spatial_layer != target_spatial_layer_ && !changing_spatial_layer_) {
```
Comment (jcague) [inline erizo/src/erizo/rtp/QualityFilterHandler.cpp]: seems like these changes come from another PR, please remove them from this one

File: erizo/src/erizo/rtp/QualityFilterHandler.h [hunk 4]
```diff
@@ -64,6 +64,9 @@ class QualityFilterHandler: public Handler, public std::enable_shared_from_this<
   time_point time_change_started_;
   int picture_id_offset_;
   int last_picture_id_sent_;
+  clock::time_point last_sent_packet = clock::now();
```
Comment (jcague) [inline erizo/src/erizo/rtp/QualityFilterHandler.h]: consider using a trailing underscore: `last_sent_packet_`

File: erizo/src/erizo/rtp/QualityFilterHandler.h [hunk 5]
```diff
@@ -64,6 +64,9 @@ class QualityFilterHandler: public Handler, public std::enable_shared_from_this<
...
+  clock::time_point cooldown_target;
```
Comment (jcague) [inline erizo/src/erizo/rtp/QualityFilterHandler.h]: consider using `cooldown_target_`

File: erizo/src/erizo/rtp/QualityFilterHandler.h [hunk 6]
```diff
@@ -64,6 +64,9 @@ class QualityFilterHandler: public Handler, public std::enable_shared_from_this<
...
+  bool cooldown = false;
```
Comment (jcague) [inline erizo/src/erizo/rtp/QualityFilterHandler.h]: consider using `cooldown_`

Comment (jcague) [reply erizo/src/erizo/rtp/QualityFilterHandler.h]: also, I prefer initializing all things in the same place, so use the constructor to initialize this variable please.

File: erizo/src/erizo/rtp/QualityFilterHandler.h (hunk 4 above)
Comment (jcague) [reply erizo/src/erizo/rtp/QualityFilterHandler.h]: also, initialize the variable in the constructor

File: erizo/src/erizo/rtp/QualityFilterHandler.cpp [hunk 7]
```diff
@@ -120,6 +127,23 @@ void QualityFilterHandler::write(Context *ctx, std::shared_ptr<DataPacket> packe
   if (is_scalable_ && !chead->isRtcp() && enabled_ && packet->type == VIDEO_PACKET) {
     RtpHeader *rtp_header = reinterpret_cast<RtpHeader*>(packet->data);
 
+    if (cooldown && clock::now() > cooldown_target) {
+    	cooldown = false;
+    	ELOG_DEBUG("End cooldown")
+    }
+
+    // Force switch to lower resolution if the packet frequency is > kSwitchDownTimeout (ms)
+    if (!cooldown && clock::now() - last_sent_packet > kSwitchDownTimeout) {
```
Comment (jcague) [inline erizo/src/erizo/rtp/QualityFilterHandler.cpp]: can we move this  if clause to `changeSpatialLayerOnKeyframeReceived()`? I think it makes more sense to have it there.

File: erizo/src/erizo/rtp/QualityFilterHandler.cpp [hunk 8]
```diff
@@ -120,6 +127,23 @@ void QualityFilterHandler::write(Context *ctx, std::shared_ptr<DataPacket> packe
...
+    	ELOG_DEBUG("Switching to lower quality");
+        future_spatial_layer_ = 0;
+        changing_spatial_layer_ = true;
+        target_temporal_layer_ = quality_manager_->getTemporalLayer();
+        time_change_started_ = clock::now() - kSwitchTimeout;	//	hack to force fast switch without keyframe
```
Comment (jcague) [inline erizo/src/erizo/rtp/QualityFilterHandler.cpp]: aren't we switching the layer here yet?

**Comment by kekkokk** [comment]
Hi @jcague, unfortunately I think I will not have time in the next future to review these PRs since my job requires me in other tasks. I opened these prs to share my "improvements", so if you consider them good and you have time, please refactor/lint at my place!

**Comment by kekkokk** [comment]
After intense debug I can say that it doesn't work as expected.
The idea is still good but needs a better implementation

**Pull request closed without merging**